### PR TITLE
WIP Longest common path

### PR DESCRIPTION
**Pull request by kriskowal**
@malandrew Take this, it is dangerous to implement Meta::thriftIDL alone.

### CODE REVIEW

File: thrift.js [hunk 1]
```diff
@@ -55,19 +56,32 @@ function Thrift(options) {
 
     assert(options, 'options required');
     assert(typeof options === 'object', 'options must be object');
-    assert(options.source || options.thriftFile,
-        'opts.source or opts.thriftFile required');
-
-    self.thriftFile = options.thriftFile ?
-        path.resolve(options.thriftFile) : null;
+    assert(options.source || options.entryPoint || options.thriftFile,
```
Comment (malandrew) [inline thrift.js]: kill `options.thriftFile`

File: thrift.js [hunk 2]
```diff
@@ -55,19 +56,32 @@ function Thrift(options) {
...
+    self.memo = options.memo || Object.create(null);
+
+    self.filename = options.entryPoint || options.thriftFile;
+    self.dirname = path.dirname(self.filename);
+    self.source = self.idls[options.entryPoint];
+    self.memo[self.filename] = self;
+
+    if (!self.source) {
+        assert.ok(options.fs, 'Must set opts.thriftFile on instantiation to resolve include paths');
+        self.filename = path.resolve(self.filename);
```
Comment (malandrew) [inline thrift.js]: `path.join` instead of `path.resolve`

File: thrift.js [hunk 3]
```diff
@@ -99,26 +113,21 @@ function Thrift(options) {
...
+    // references.
+    self.compile();
+    // We only link from the root Thrift object.
+    if (!options.noLink) {
         self.link();
     }
+
 }
 
 Thrift.loadSync = function loadSync(options) {
```
Comment (malandrew) [inline thrift.js]: kill `loadSync`

File: thrift.js [hunk 4]
```diff
@@ -190,14 +213,9 @@ Thrift.prototype._compile = function _compile(defs) {
...
 
-    assert(
-        self.dirname,
-        'Must set opts.thriftFile on instantiation to resolve include paths'
-    );
-
     if (def.id.lastIndexOf('./', 0) === 0 ||
         def.id.lastIndexOf('../', 0) === 0) {
-        var thriftFile = path.resolve(self.dirname, def.id);
+        var filename = path.resolve(self.dirname, def.id);
```
Comment (malandrew) [inline thrift.js]: `path.join` instead of `path.resolve`

**Comment by kriskowal** [comment]
Replacing with #95

**Pull request closed without merging**